activerecord/lib/active_record/autosave_association.rb

```diff
@@ -390,11 +390,8 @@ def save_has_one_association(reflection)
           elsif autosave != false
             key = reflection.options[:primary_key] ? send(reflection.options[:primary_key]) : id
 
-            if (autosave && record.changed_for_autosave?) || new_record? || record_changed?(reflection, record, key)
-              unless reflection.through_reflection
-                record[reflection.foreign_key] = key
-              end
-
+            if (autosave && record.changed_for_autosave?) || (!reflection.through_reflection && record_changed?(reflection, record, key))
```

Comment:
I see now, I think we can safely remove this indeed because of the method https://github.com/rails/rails/blob/62cc6d887b2f86be407167e0d2f19653231748c0/activerecord/lib/active_record/autosave_association.rb#L447 also returns true if it is a new record. 
However the name of that method seems a bit misleading. I would expect that method to not care about `new_record` and have it the `new_record` check back in this IF. 

ref https://github.com/aghull/rails/commit/8022fc4913d5fa285889795617a1f37c5aa705a9
cc @laurocaetano